crates/noirc_frontend/src/lexer/token.rs

```diff
@@ -321,28 +321,42 @@ impl IntType {
 pub enum TestScope {
     /// If a test has a scope of ShouldFail, then it is expected to fail
     ShouldFail,
+    /// If a test has a scope of ShouldFailWith, then it can only pass
+    /// if it fails with the specified reason.
+    ShouldFailWith(String),
     /// No scope is applied and so the test must pass
     None,
 }
 
 impl TestScope {
     fn lookup_str(string: &str) -> Option<TestScope> {
```

Comment:
Attributes parsing is becoming a small/midly complicated FSA, so it makes sense to put this into its own file and then have tests just for attribute parsing. 

WIll be a bit janky because the attribute parser does not receive `#[test(should_fail)]` from the lexer but `test(should_fail)`

Reply:
Attributes could be fed into the actual parser instead. We could have an `attribute` parser to parse `'#' '[' contents ']'` with the contents being an identifier or function call (allowing for `a = b` arguments)

Reply:
thats a good shout